MTM-61167 , OTLP instrumentation for microservices (#3742)
SoftwareAG/c8y-docs: content/microservice-sdk/general-aspects-bundle/microservice-otlp-monitoring.md

The Java Microservice SDK leverages the [OpenTelemetry Java agent](https://opentelemetry.io/docs/zero-code/java/agent/) which provides automatic instrumentation for many popular libraries and frameworks.

### Configuration
Microservices can obtain the OTLP configuration from their individual tenant options. 
The tenant option category containing the OTLP parameters is determined in the following order:

- The settingsCategory defined in the microservice manifest
- The microservice’s context path
- The microservice name

A detailed list of configuration parameters can be found in the [OTLP configuration documentation](https://opentelemetry.io/docs/languages/java/configuration/).

#### OTLP tenant options
Tenant options with OTLP configuration parameters can be set via [REST commands](https://cumulocity.com/api/core/2025/#tag/Options) sent to 
the tenant options endpoint of the tenant.
For example, setting the endpoint where OTLP signals are exported to is done with this json document:

```json
{
"category": "<application-name>",
"key": "otel.exporter.otlp.endpoint",
"value": "https://otlp-gateway.net/otlp"
}

> ```

Reviewer: Could you add information (sub-chapter?) about security? It is important that users understand how to store credentials in secure way.

Author: chapter added

As an example, the parameter values required to export OTLP signals to Grafana (without using an OTLP collector instance) could be set like this:

```properties
otel.exporter.otlp.endpoint: https://otlp-gateway-prod.grafana.net/otlp
otel.exporter.otlp.protocol: http/protobuf

> otel.exporter.otlp.headers: Authorization=Basic MTAxNTI0OTpnbGNfZ…

Reviewer: Sorry but I am confused. One of the key advantages of using tenant options over static approaches (like configuring OTLP through the application.yaml file) is proper handling of credentials to authenticate against the remote monitoring system. However this example shows the authorization header in 'plain' text (it actually displays the header in Base64 encoding, which is not encryption). I think this example is quite misleading.

Author: changed to credentials.otel.exporter.otlp.headers ...

```

> The `otel.service.name` parameter value is automatically set with the microservice pod name at run time.

Reviewer: The documentation doesn't make it clear where these properties live. User might think of Core properties(?) It would be better to make it clear that the properties stored in the tenant options are mapped 1:1 on _environment parameters_ of the microservice.

Author: chapter about conversion to env.variables added above

The parameter `otel.resource.attributes` gets automatically assigned these properties:

```properties
service.name: application name
service.instance.id: microservice pod name
service.namespace: tenant name
service.version: application version
```

OTEL parameters are only set in the microservice deployment if either 
`otel.sdk.disabled=false` or `otel.javaagent.enabled=true`.
By default, OTLP parameters in the deployment are:
```properties
otel.sdk.disabled=true
otel.javaagent.enabled=false
```

{{< c8y-admon-important >}}
To let parameter changes take effect, the microservice must be unsubscribed and subscribed again.
{{< /c8y-admon-important >}}

### Enabling auto-instrumentation
Whether the microservice application gets instrumented by the java-agent is controlled 
with the parameter `otel.javaagent.enabled`. Setting it to `true` enables the instrumentation:

```json
{
"category": "<application-name>",
"key": "otel.javaagent.enabled",
"value": "true"
}
```

> If enabled, the Java agent JAR file file will be downladed and injected into the microservice JVM at startup time.

Reviewer: It is important to make people understand that enabling or disabling requires each unsubscribe and resubscribe operation. Could you please add a note?

Author: see a few lines above:

{{< c8y-admon-important >}}
To let parameter changes take effect, the microservice must be unsubscribed and subscribed again.
{{< /c8y-admon-important >}}

> Configuring auto-instrumentation for only selected libraries or frameworks or allowing 

Reviewer: The sentence is a bit confusing in its current form - at least for me. I would suggest rephrasing it this way
```suggestion
Configuring auto-instrumentation for selected libraries or frameworks, or opting for manual instrumentation only, is described in the [Opentelemetry instrumention documentation](https://opentelemetry.io/docs/zero-code/java/agent/disable/).
```

Author: copied manually to avoid merge conflicts

---
weight: 95
title: Microservice monitoring
layout: redirect
---

{{< product-c8y-iot >}} supports the [OpenTelemetry framework](https://opentelemetry.io/) for exporting telemetry data (metrics, logs, and traces) from your microservice to help you analyze your application’s performance and behavior. 
The Java Microservice SDK leverages the [OpenTelemetry Java agent](https://opentelemetry.io/docs/zero-code/java/agent/) which provides automatic instrumentation for many popular libraries and frameworks.

> 

Reviewer: I think we need an introduction chapter explaining why tis feature is so important for microservice providers

- current situation and limitatations: All log data is sent to the standard output and persisted temporarily by the infrastructure. You can access log data via Rest interface but limited to last 35 MB. Logs are lost in case of a restart. Server runtime metrics like memory and CPU consumption are not directly (if not written to the log by the application)
- Ability to monitor application behavior at runtime is of crucial importance wrt relialibity and performance of the microservice application
- Altough log aggregation was possible already (see https://community.cumulocity.com/t/log-aggregation-for-cumulocity-iot-microservices/7477) you had to program your microservices against a specific technology like Grafana. OpenTelemetry removes this vendor-lock-in. It defines a standard for application metrics, logging, traces which doesn't depend on which technology is used as remote monitring system

Author: Addtl. text included now, with a few modifications; as discussed

> ### Configuration

Reviewer 2: ```suggestion
### Configuration {#configuration}
```

Reviewer 2: Please add the heading IDs to the other headings in this file in the same way.

Author: Done

> ### Configuration {#configuration}

Reviewer 2: It seems you missed my comment that these heading IDs must be added to all headings in this file.

Author: Sorry, tags are included now.

> Microservices can obtain the OTLP configuration from their individual tenant options. 

Reviewer 3: I would add "current" or "existing" OTLP configuration

Author: added "current"

> The tenant option category containing the OTLP parameters is determined in the following order:

Reviewer 3: Can we add an example how to obtain the current configuration in this context.

Author: there are 2 links to the Options sections of the OpenApi documentation

- The settingsCategory defined in the microservice manifest
- The microservice’s context path
- The microservice name

> A detailed list of configuration parameters can be found in the [OTLP configuration documentation](https://opentelemetry.io/docs/languages/java/configuration/).

Reviewer 3: I would add this sentence to the next section where you say that the configuration parameters can be set via Rest calls

Author: sentence moved

#### OTLP tenant options

> Tenant options with OTLP configuration parameters can be set via [REST commands](https://cumulocity.com/api/core/2025/#tag/Options) sent to 

Reviewer 3: Maybe put an emphasis that the customer can change the existing OTLP configuration by saying sth like: 
"The existing OTLP configuration can be changed via tenant options"..

Author: why? It would be quite unexpected if the tenant options could not be changed.

the tenant options endpoint of the tenant.

> For example, setting the endpoint where OTLP signals are exported to is done with this json document:

Reviewer 2: ```suggestion
For example, setting the endpoint to which OTLP signals are exported is done with this JSON document:
```

```json
{
"category": "<application-name>",
"key": "otel.exporter.otlp.endpoint",
"value": "https://otlp-gateway.net/otlp"
}
```

> The Opentelemetry parameter names in this document are according to the specification in [OTLP configuration documentation](https://opentelemetry.io/docs/languages/java/configuration/). 

Reviewer 2: ```suggestion
The OpenTelemetry parameter names in this document follow the specification in the [OTLP configuration documentation](https://opentelemetry.io/docs/languages/java/configuration/). 
```

The parameters are stored as tenant options and injected into the microservice as environment variables.
To convert a tenant option to an environment variable, these steps are applied as described [here](https://opentelemetry.io/docs/languages/java/configuration/#environment-variables-and-system-properties):
- Convert the name to uppercase.
- Replace all `.` and `-` characters with `_`.

For example, the `otel.sdk.disabled` tenant option is equivalent to the `OTEL_SDK_DISABLED` environment variable.

##### Encryption
Tenant options containing values to be encrypted, like passwords or access tokens, must be preceded with a `credentials.` prefix, 

> as described in the [Encryption](/microservice-sdk/general-aspects/#encryption) chapter.

Reviewer 2: ```suggestion
as described in the [Encryption](/microservice-sdk/general-aspects/#encryption) section.
```

}
```

##### OTEL parameter values

As an example, the parameter values required to export OTLP signals to Grafana (without using an OTLP collector instance) could be set like this:

```properties
otel.exporter.otlp.endpoint: https://otlp-gateway-prod.grafana.net/otlp
otel.exporter.otlp.protocol: http/protobuf
credentials.otel.exporter.otlp.headers: <authentication data>
```

The `otel.service.name` parameter value is automatically set with the microservice pod name at run time.

The parameter `otel.resource.attributes` gets automatically assigned these properties:

```properties
service.name: application name
service.instance.id: microservice pod name
service.namespace: tenant name
service.version: application version
```

OTEL parameters are only set in the microservice deployment if either 
`otel.sdk.disabled=false` or `otel.javaagent.enabled=true`.
By default, OTLP parameters in the deployment are:
```properties
otel.sdk.disabled=true
otel.javaagent.enabled=false
```

{{< c8y-admon-important >}}
To let parameter changes take effect, the microservice must be unsubscribed and subscribed again.
{{< /c8y-admon-important >}}

### Enabling auto-instrumentation

> Whether the microservice application gets instrumented by the java-agent is controlled 

Reviewer 2: ```suggestion
Whether the microservice application gets instrumented by the Java agent is controlled 
```

> Whether the microservice application gets instrumented by the Java agent is controlled 

Reviewer: Minor, feel free to ignore if you don't like the suggestion
```suggestion
Whether the microservice application gets instrumented by the OpenTelemetry Java agent is controlled 
```

with the parameter `otel.javaagent.enabled`. Setting it to `true` enables the instrumentation:

```json
{
"category": "<application-name>",
"key": "otel.javaagent.enabled",
"value": "true"
}
```

> If enabled, the Java agent JAR file file will be downladed and attached to the microservice JVM at startup time.

Reviewer 2: ```suggestion
If enabled, the Java agent JAR file is downloaded and attached to the microservice JVM at startup time.
```

Configuring auto-instrumentation for selected libraries or frameworks, or opting for manual instrumentation only, is described 

> in the [Opentelemetry instrumention documentation](https://opentelemetry.io/docs/zero-code/java/agent/disable/).

Reviewer 2: ```suggestion
in the [OpenTelemetry instrumention documentation](https://opentelemetry.io/docs/zero-code/java/agent/disable/).
```

### Manual instrumentation

> In parallel to the automatic instrumentation, manual instrumentation of the microservice application is possible as well.

Reviewer 2: ```suggestion
In parallel with the automatic instrumentation, manual instrumentation of the microservice application is also possible.
```

> In parallel with the automatic instrumentation, manual instrumentation of the microservice application is also possible.

Reviewer: I am not a native speaker but that sounds clunky. At least AI states that...:-)
```suggestion
In addition to automatic instrumentation, microservices can be manually instrumented.
```

The Java agent creates the GlobalOpenTelemetry object which can be used as a starting point to create 
individual Tracer or Meter objects for [custom instrumentation](https://opentelemetry.io/docs/zero-code/java/agent/api/).

Java code example:
```java
import io.opentelemetry.api.GlobalOpenTelemetry;
import io.opentelemetry.api.metrics.Meter;

Meter meter = GlobalOpenTelemetry.getMeter("application");
```

> A basic example for this use case can be found [here](https://github.com/open-telemetry/opentelemetry-java-examples/tree/main/javaagent/src/main/java/io/opentelemetry/example/javagent).

Reviewer 2: ```suggestion
A basic example for this use case can be found in the [OpenTelemetry GitHub repository](https://github.com/open-telemetry/opentelemetry-java-examples/tree/main/javaagent/src/main/java/io/opentelemetry/example/javagent).
```

If instrumentation with the Java agent is disabled, complete manual instrumentation without the Java agent can be applied as well.

> Detailed examples for various use cases can be found [here](https://github.com/open-telemetry/opentelemetry-java-examples/tree/main).

Reviewer 2: ```suggestion
Detailed examples for various use cases can be found in the [OpenTelemetry GitHub repository](https://github.com/open-telemetry/opentelemetry-java-examples/tree/main).
```

---
weight: 95
title: Microservice monitoring
layout: redirect
---

Monitoring microservice behavior at runtime is crucial for ensuring the stability, reliability, and performance of the application.

> Monitoring of microservices on the {{< product-c8y-iot >}} platform is already possible to some extent, however with additional effort and some limitations:

Reviewer: Proposal: replace lines 8-16 with the following content:

{{< product-c8y-iot >}} allows you to export monitoring data to a customer-provided observability backend using OpenTelemetry.

Reviewer 2: Yes, agreed. Will there be a respective change log entry?

Author: Done: replaced lines 8 - 16

Basically, the option to use OTLP is an additional feature. There is no change to existing functionality.
Do we have change log entries even is this case?
There will be an technical article describing the advantages of using OTLP.

- All logging data is sent to the standard output and persisted temporarily by the infrastructure. 
You can access logging data via the Rest interface but limited to the last 35 MB. Logs are lost in case of a restart. 

- Although [log aggregation](https://community.cumulocity.com/t/log-aggregation-for-cumulocity-iot-microservices/7477) is possible already, 
you have to program your microservices against a specific technology like Grafana. 

- Server runtime metrics like memory and CPU consumption are not directly logged (if not written to the log by the application).

The OpenTelemetry standard provides a holistic approach to collect and export application logs, metrics, and traces to monitoring systems of various vendors.

> You can freely define which data is to be collected and to which endpoint it is to be sent to.

Reviewer 2: ```suggestion
You can freely define which data is collected and to which endpoint it is sent.
```

---
weight: 95
title: Microservice monitoring
layout: redirect
---

> ### Current situation and limitatations {#current-situation-and-limitatations}

Reviewer 2: @mgrumann If this is the current situation, is the new approach still in preview? Otherwise, this would be the previous situation, right? 
Actually, we should be careful with these before/after statements as they become outdated after a while. Will this section be removed with the next release when the new situation is the standard?

Reviewer: Agreed. Robert and I just talked about this point. We ae will simply omit this header altogether as not required

Author: The header is removed now.

---
weight: 95
title: Microservice monitoring
layout: redirect
---

### Current situation and limitatations {#current-situation-and-limitatations}

> The ability to monitor microservice behavior at runtime is of crucial importance with respect to stability, relialibity, and performance of the microservice application.

Reviewer: ```suggestion
Monitoring microservice behavior at runtime is crucial for ensuring the stability, reliability, and performance of the application.
```

Monitoring of microservices on the {{< product-c8y-iot >}} platform is already possible to some extent, however with additional effort and some limitations:

- Server runtime metrics like memory and CPU consumption are not directly logged (if not written to the log by the application).

> - All logging data is sent to the standard output and persisted temporarily by the infrastructure. 

Reviewer: This point shall be moved to first position in this bullet item list since it is the most important point.

Author: Bullet points are reordered now

You can access logging data via the Rest interface but limited to the last 35 MB. Logs are lost in case of a restart. 

- Although [log aggregation](https://community.cumulocity.com/t/log-aggregation-for-cumulocity-iot-microservices/7477) is possible already, 
you have to program your microservices against a specific technology like Grafana. 

The OpenTelemetry standard provides a holistic approach to collect and export application logs, metrics, and traces to monitoring systems of various vendors.

> The customer can freely define which data is to be collected and to which endpoint it is to be sent to.

Reviewer: ```suggestion
You can freely define which data is to be collected and to which endpoint it is to be sent to.
```

OpenTelemetry is an open standard with implementations supporting several [languages](https://opentelemetry.io/docs/languages/) like Java, Python, Go, Ruby, C++.
{{< product-c8y-iot >}} provides the option of zero-code instrumentation for microservices developed with the Microservice SDK for Java. 

> The so-called OpenTelemetry Java agent JAR file gets attached to the JVM, and the only configuration data needed is the access information of the monitoring system. 

Reviewer: I suggest to remove this sentence entirely. This is a technical detail how this works fro Java and should be given in linked section if not done yet
```suggestion
```

> Detailed instructions can be found in the [Microservice SDK for Java](/microservice-sdk/java/#otlp-configuration) section.

Reviewer: ```suggestion
Refer to the [Microservice SDK for Java](/microservice-sdk/java/#otlp-configuration) section for details.
```
